packages/gatsby/src/utils/webpack/plugins/static-query-mapper.ts

```diff
@@ -83,293 +83,293 @@ export class StaticQueryMapper {
+    //           realPathCache,
+    //           staticQuery.componentPath
+    //         )
+
+    //         if (
+    //           !doesModuleMatchResourcePath(
+    //             staticQueryComponentPath,
+    //             webpackModule
+    //           )
+    //         ) {
+    //           continue
+    //         }
+
+    //         let set = webpackModulesByStaticQueryId.get(staticQuery.hash)
+
+    //         if (!set) {
+    //           set = new Set()
+    //           webpackModulesByStaticQueryId.set(staticQuery.hash, set)
+    //         }
+
+    //         set.add(webpackModule)
+    //       }
+
+    //       for (const [filePath, slices] of componentsUsingSlices) {
+    //         const componentComponentPath = getRealPath(realPathCache, filePath)
+
+    //         if (
+    //           !doesModuleMatchResourcePath(
+    //             componentComponentPath,
+    //             webpackModule
+    //           )
+    //         ) {
+    //           continue
+    //         }
+
+    //         webpackModulesByUsedSlicePlaceholderAlias.set(webpackModule, slices)
+    //       }
+
+    //       for (const component of components.values()) {
+    //         const componentComponentPath = getRealPath(
+    //           realPathCache,
+    //           component.componentPath
+    //         )
+
+    //         if (
+    //           !doesModuleMatchResourcePath(
+    //             componentComponentPath,
+    //             webpackModule
+    //           )
+    //         ) {
+    //           continue
+    //         }
+
+    //         componentModules.set(webpackModule, component)
+    //       }
+    //     }
+
+    //     function traverseModule(
+    //       module: NormalModule,
+    //       config: {
+    //         onComponent(component: IGatsbyPageComponent): void
+    //         onRoot(): void
+    //       },
+    //       visitedModules = new Set<NormalModule>()
+    //     ): void {
+    //       if (visitedModules.has(module)) {
+    //         return
+    //       }
+    //       visitedModules.add(module)
+
+    //       if (module === asyncRequiresModule) {
+    //         return
+    //       }
+
+    //       const component = componentModules.get(module)
+    //       if (component) {
+    //         config.onComponent(component)
+    //         // don't return here yet, as component might be imported by another one, and we want to traverse up until we reach async-requires
+    //       }
+
+    //       if (entryModules.has(module)) {
+    //         config.onRoot()
+    //         return
+    //       }
+
+    //       const incomingConnections =
+    //         compilation.moduleGraph.getIncomingConnections(module)
+    //       for (const connection of incomingConnections) {
+    //         if (connection.originModule instanceof NormalModule) {
+    //           traverseModule(connection.originModule, config, visitedModules)
+    //         }
+    //       }
+    //     }
```

Comment:
maybe we can figure out alternative way to "backtrack" imports that is not relying on `moduleGraph.getIncomingConnections`?

Or if we zoom out of this implementation detail, what could be other way to determine if module A is used/needed/imported by module B?

There is additional consideration to make here:
For React Server Components support (experimental, not on by default) we have to do this moduleGraph traversal BEFORE we mess with module graph to make React Server Components work by removing non-client components and add entries for client components, so that's something to keep in mind that we likely can't just use final `Stats` object for that (while it probably be fine for Rspack, as it would be ok to not support Server Components there yet, this would introduce separate code path / implementation that adds too much maintainance burden to reallistically support and possibly introduce behavior changes between the 2 implemantions)